nucypher/blockchain/eth/currency.py

```diff
@@ -0,0 +1,197 @@
+from _pydecimal import Decimal
+
+import eth_utils
+import maya
+from eth_utils import currency
+from nacl.hash import sha256
+from typing import Union, Tuple
+
+from nucypher.blockchain.eth.agents import NucypherTokenAgent
+from nucypher.blockchain.eth.constants import SUBDIGITS
+from nucypher.blockchain.eth.utils import datetime_at_period, datetime_to_period
+
+
+class NU:
+    """
+    An amount of NuCypher tokens that doesn't hurt your eyes.
+    Wraps the eth_utils currency conversion methods.
+
+    The easiest way to use NU, is to pass an int, float, or str, and denomination string:
+
+    Int:    nu = NU(100, 'NU')
+    Int:    nu_wei = NU(15000000000000000000000, 'NUWei')
+
+    Float:  nu = NU(15042.445, 'NU')
+    String: nu = NU('10002.302', 'NU')
+
+    ...or alternately...
+
+    Float: nu = NU.from_tokens(100.50)
+    Int: nu_wei = NU.from_nu_wei(15000000000000000000000)
+
+    Token quantity is stored internally as an int in the smallest denomination,
+    and all arithmetic operations use this value.
+    """
+
+    __symbol = 'NU'
+    __decimals = SUBDIGITS
+    __agent_class = NucypherTokenAgent
+
+    # conversions
+    __denominations = {'NUWei': 'wei',
+                       'NU': 'ether'}
+
+    def __init__(self, value: Union[int, float, str], denomination: str):
+
+        # Calculate smallest denomination and store it
+        wrapped_denom = self.__denominations[denomination]
+
+        # Validate Early
+        if '.' in str(value):
+
+            _, fraction = str(value).split('.')
+            if len(fraction) > self.__decimals:
+                raise ValueError("Cannot initialize with fractional wei value")
+
+            if wrapped_denom == 'wei':
+                raise ValueError("Cannot initialize with fractional wei value")
+
+        self.__value = currency.to_wei(number=value, unit=wrapped_denom)
+
+    @classmethod
+    def from_nu_wei(cls, value: int):
+        return cls(value, denomination='NUWei')
+
+    @classmethod
+    def from_tokens(cls, value: Union[int, float, str]):
+        return cls(value, denomination='NU')
+
+    def to_tokens(self) -> Decimal:
+        """Returns an decimal value of NU"""
+        return currency.from_wei(self.__value, unit='ether')
+
+    def to_nu_wei(self) -> int:
+        """Returns an int value in NU-Wei"""
+        return int(self.__value)
+
+    def __eq__(self, other) -> bool:
+        return int(self) == int(other)
+
+    def __radd__(self, other) -> int:
+        return int(self) + other
```

Comment:
I think `__radd__` and `__rmul__` should behave exactly as `__add__` and `__mul__` respectively. For example, if we have an `amount` in NU and we compute `2 * amount`, I'd expect the result is still `NU`. What's the motivation for returning `int` here?

Reply:
I had a misunderstanding of these dunder methods.  I'll make an update.

Reply:
:heavy_check_mark: